velox/exec/ExchangeQueue.cpp

```diff
@@ -91,29 +91,42 @@ void ExchangeQueue::enqueueLocked(
   }
 }
 
-std::unique_ptr<SerializedPage> ExchangeQueue::dequeueLocked(
+std::vector<std::unique_ptr<SerializedPage>> ExchangeQueue::dequeueLocked(
+    uint32_t maxBytes,
     bool* atEnd,
     ContinueFuture* future) {
   VELOX_CHECK(future);
   if (!error_.empty()) {
     *atEnd = true;
     VELOX_FAIL(error_);
   }
-  if (queue_.empty()) {
-    if (atEnd_) {
-      *atEnd = true;
-    } else {
-      promises_.emplace_back("ExchangeQueue::dequeue");
-      *future = promises_.back().getSemiFuture();
-      *atEnd = false;
+
+  std::vector<std::unique_ptr<SerializedPage>> pages;
+  uint32_t pageBytes = 0;
+  for (;;) {
+    if (queue_.empty()) {
+      if (atEnd_) {
+        *atEnd = true;
+      } else {
+        promises_.emplace_back("ExchangeQueue::dequeue");
+        *future = promises_.back().getSemiFuture();
+        *atEnd = false;
+      }
+      return pages;
+    }
+
+    if (pageBytes > 0 && pageBytes + queue_.front()->size() > maxBytes) {
+      return pages;
     }
-    return nullptr;
+
+    pages.emplace_back(std::move(queue_.front()));
+    queue_.pop_front();
+    *atEnd = false;
```

Comment:
nit: shall we just set atEnd = false once at the beginning and set to true at l109? Thanks!

Reply:
Sure. Will change.